src/selectors/modular-pipelines.js

```diff
@@ -0,0 +1,54 @@
+import { createSelector } from 'reselect';
+import { getPipelineModularPipelineIDs } from './pipeline';
+
+const getNodesModularPipelines = (state) => state.node.modularPipelines;
+const getModularPipelineIDs = (state) => state.modularPipeline.ids;
+const getModularPipelineName = (state) => state.modularPipeline.name;
+const getModularPipelineEnabled = (state) => state.modularPipeline.enabled;
+
+/**
+ * Retrieve the formatted list of modular pipeline filters
+ */
+export const getModularPipelineData = createSelector(
+  [getModularPipelineIDs, getModularPipelineName, getModularPipelineEnabled],
+  (modularPipelineIDs, modularPipelineName, modularPipelineEnabled) =>
+    modularPipelineIDs.sort().map((id) => ({
+      id,
+      name: modularPipelineName[id],
+      enabled: Boolean(modularPipelineEnabled[id]),
+    }))
+);
+
+/**
+ * Get the total and enabled number of modular pipelines
+ */
+export const getModularPipelineCount = createSelector(
+  [getPipelineModularPipelineIDs, getModularPipelineEnabled],
+  (modularPipelineIDs, modularPipelineEnabled) => ({
+    total: modularPipelineIDs.length,
+    enabled: modularPipelineIDs.filter((id) => modularPipelineEnabled[id])
+      .length,
+  })
+);
+
+/**
+ * returns an array of modular pipelines with the corresponding
+ * nodes for each modular pipeline
+ */
+export const getModularPipelineNodes = createSelector(
+  [getNodesModularPipelines, getModularPipelineIDs],
+  (allNodes, modularPipelines) => {
+    return modularPipelines.map((modularPipeline) => {
+      let nodes = [];
+      Object.keys(allNodes).map((key) =>
```

Comment:
I think since the return result of `map` isn't used here and the loop uses side effects it might be best to convert this into a `for` loop?

Reply:
Thanks for the comment, though I am not entirely sure of the advantages of converting this to a `for` loop given that in our codebase ( and generally for most JS development nowadays) we often favor using `map`?  This is a pttern that is used in our other selectors as well. 

personally I also think using a `map` allows the code to be much cleaner.

Reply:
I commented on this one too below and suggested using `forEach`. `map` is typically used to assign the result of a loop to a variable (e.g. `const result = array.map(d => d.foo)`), which this bit of code doesn't do. Hence, using `forEach` or `for` is clearer to a reader, as it doesn't imply that you intended to return the iterated array.

Reply:
Yeah `map` and friends return an array themselves already. You're right it's a sound idea in this case to convert this to a functional style though instead, but what you need here looks like `filter` rather than `map`  e.g. something like:

```js
  const nodes = Object.keys(allNodes)
        .filter(key => allNodes[key] && allNodes[key].includes(modularPipeline));
```